### DISCUSSION TITLE
In-place register filters

### DISCUSSION THREAD

**Discussion by allan-on** [opensrp/fhircore#2138] [Ideas]
### Context 
The in-place filtering is a toggle filter that filters the register based on any or all of the supported attributes
This would provide advanced filtering that's not supported by search from the search bar or by filtering of registers by passing the filter params as implemented in #2036  

#### Example
Consider we have a register of patients screened for risk of hypertension. 
It should be possible to filter the register based on the Patient condition such as _low-risk_, _high-risk_, Patients with followups   e.t.c

### Requirements

1. The filters should filter/work on the register being displayed/viewed without having to navigate away
2. The filters should be configurable - supplied via configs
3. It should support filtering by any of the rules defined for the register
4. We should probably define an appropriate UI/UX that provides toggling/activating of the different filters

### Additional context
Resource filtering implemented/supported
- [Register configuration](https://fhircore.smartregister.org/configurations/register-config)
- #2036 
- #2001

**Comment by pld** [comment]
Cool, yea I think the UX is a big part of this, maybe we should work on mock-ups for that first, then we can circle back to the tech details? We need the UX clear so that it drives the tech

**Reply by allan-on** [reply]
I agree. Thanks

**Reply by HenryRae** [reply]
@pld @f-odhiambo @Gental-Giant @AnnieMungai @dubdabasoduba 
Are there any other requirements or use cases we may want to consider for filters?

Some of the other use cases am thinking of:
1. on the home register, allow user to filter the list but due/overdue count, specific care plans(list families with patients having specific care plans). 
2. allow the user to filter by more than one option
3. allow user sort filtered results

cc: @rowo

**Reply by pld** [reply]
yes all of those are valid

**Comment by allan-on** [comment]
Hi @rowo @HenryRae We need  help with mockups for this implementation

**Comment by allan-on** [comment]
A similar feature implementation in Malawi Core is documented in [Appointment Register Filters #2102 ](https://github.com/opensrp/fhircore/discussions/2102)
TL;DR
- The (Appointment) Register can be filtered by the options displayed on the Filter dialog such as Filtering by _Patient categories_ or _Appointment Reasons_ etc

<img src="https://user-images.githubusercontent.com/16164649/225630679-fae351a9-975a-4d0b-ba8a-468aa467ee50.png" height=600 width=300 />

<img src="https://user-images.githubusercontent.com/16164649/225631196-7dd0b1c7-feb5-4380-aeef-13f04bfabd6a.png" height=600 width=300 />

**Comment by allan-on** [comment]
For the first version of the implementation, we'll work with the following design: 

![register_filters_1 0_annotated](https://github.com/opensrp/fhircore/assets/16164649/86900968-807f-4e7d-b18e-1fca5fd8553f)

### Filter UI

a. The register filters will be defined in a filter config specifying the enabling of the feature and other required filter parameters such that **1** is only visible when filtering is enabled. 
b. The filter (options) form will be defined as a Questionnaire for reconfigurability and reuse. The filter icon **1** launches this filter options Questionnaire view **2**. 
c. The resulting count can be updated and shown when the filter is applied  in **3**

### Filtering 
a. We'll add register filter configurations as part of the register config. This register filter configuration will include the reference to the filter options `Questionnaire` (and the base resource on which the filter is applied?).   
b. When the `Questionnaire` is submitted, the filter options can be read from the resulting `QuestionnaireResponse`. This `QuestionnaireResponse` will not be persisted in the DB.  
c. The retrieved options would then be used to build the parameters for the filter operation that would then populate the data map used when rendering the register. 

----

The above approach works when filtering using attributes of the base resource that's used in that register. e.g. For a Patient register filter by gender, is active, city etc. 

Using the example in the design mocks, `Patient` is the base resource while `Tasks` are the related resource in the Patient/Client register. 
To filter clients by `Task` status, the `Task` resource should be the base resource then include `Patient` as a related resource.

**Reply by dubdabasoduba** [reply]
@allan-on how are we mapping the QR items to Resource search Parameters?

**Comment by HenryRae** [comment]
@allan-on in this filter questionnaire, will we be able to have:
1. a different header color from default questionnaires(default questionnaires are using white as I presume)?
<img width="175" alt="Screenshot 2023-08-02 at 11 18 57" src="https://github.com/opensrp/fhircore/assets/31955469/d8c299d3-dd13-444f-ba25-7d72602806c5">

2. a clear action on the header? we don't have any button CTAs on the right side of the questionnaires. 
<img width="94" alt="Screenshot 2023-08-02 at 11 19 07" src="https://github.com/opensrp/fhircore/assets/31955469/d6e25905-2de1-492c-9013-d0ccf0ca393f">

Are the questionnaires configurable to that extent i.e see attached

**Reply by dubdabasoduba** [reply]
@HenryRae @allan-on I might be wrong but I am not sure we can achieve the Clear All without re-rendering the questionnaire. This is something we need to review.

**Reply by HenryRae** [reply]
please keep us posted on any findings
